src/analysis/souffle/taint.dl

```diff
@@ -113,10 +116,21 @@ lacksIntegrityTag(dst, owner, integTag) :-
   edge(src, dst), lacksIntegrityTag(src, owner, integTag),
   !hasAppliedIntegrityTag(dst, owner, integTag).
 
-// An access path must have an integrity tag if it does not lack the integrity tag.
+// Any AccessPath that has an IntegrityTag directly applied to it may (really,
+// stronger than may, must) have that IntegrityTag.
+mayHaveIntegrityTag(ap, owner, integTag) :- hasAppliedIntegrityTag(ap, owner, integTag).
+
+// Any AccessPath that has an incoming resolvedEdge where the source of that
+// edge may have an IntegrityTag may also have that IntegrityTag.
+mayHaveIntegrityTag(dst, owner, integTag) :-
+  resolvedEdge(owner, src, dst), mayHaveIntegrityTag(src, owner, integTag),
+  !removeIntegrityTag(dst, owner, integTag).
+
+// An access path must have an integrity tag if it may have the IntegrityTag
+// and does not lack the IntegrityTag
 mustHaveIntegrityTag(ap, owner, integTag) :-
   isAccessPath(ap), isIntegrityTag(integTag), isPrincipal(owner),
-  !lacksIntegrityTag(ap, owner, integTag).
+  mayHaveIntegrityTag(ap, owner, integTag), !lacksIntegrityTag(ap, owner, integTag).
```

Comment:
In this rule `isIntegrityTag` is only true for the seed tags: `TimestampMillis` and `Constant_1000`.  Specifically, `isIntegrityTag` is not true for `TimestampSeconds`, and `TimestampRedacted`.

Reply:
Good point. I've added explicit `isIntegrityTag` facts for the `IntegrityTag`s added by policy rules.